pkg/listwatch/namespace_denylist.go

```diff
@@ -0,0 +1,159 @@
+
+package listwatch
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// denylistListerWatcher implements cache.ListerWatcher
+// which wraps a cache.ListerWatcher,
+// filtering list results and watch events by denied namespaces.
+type denylistListerWatcher struct {
+	denylist map[string]struct{}
+	next     cache.ListerWatcher
+	logger   log.Logger
+}
+
+// newDenylistListerWatcher creates a cache.ListerWatcher
+// wrapping the given next cache.ListerWatcher
+// filtering lists and watch events by the given namespaces.
+func newDenylistListerWatcher(l log.Logger, namespaces []string, next cache.ListerWatcher) cache.ListerWatcher {
+	denylist := make(map[string]struct{})
+
+	for _, ns := range namespaces {
+		denylist[ns] = struct{}{}
+	}
+
+	return &denylistListerWatcher{
+		denylist: denylist,
+		next:     next,
+		logger:   l,
+	}
+}
+
+func (w *denylistListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	l := metav1.List{}
+
+	list, err := w.next.List(options)
+	if err != nil {
+		return nil, err
+	}
+
+	objs, err := meta.ExtractList(list)
+	if err != nil {
+		return nil, err
+	}
+
+	metaObj, err := meta.ListAccessor(list)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, obj := range objs {
+		acc, err := meta.Accessor(obj)
+		if err != nil {
+			return nil, err
+		}
+
+		logger := log.With(w.logger, "name", fmt.Sprintf("%v/%v", acc.GetNamespace(), acc.GetName()))
+
+		if _, denied := w.denylist[acc.GetNamespace()]; denied {
+			level.Debug(logger).Log("msg", "denied")
+			continue
+		}
+
+		level.Debug(logger).Log("msg", "allowed")
+
+		l.Items = append(l.Items, runtime.RawExtension{Object: obj.DeepCopyObject()})
+	}
+
+	l.ListMeta.ResourceVersion = metaObj.GetResourceVersion()
+	return &l, nil
+}
+
+func (w *denylistListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	nextWatch, err := w.next.Watch(options)
+	if err != nil {
+		return nil, err
+	}
+
+	return newDenylistWatch(w.logger, w.denylist, nextWatch), nil
+}
+
+// newDenylistWatch creates a new watch.Interface,
+// wrapping the given next watcher,
+// and filtering watch events by the given namespaces.
+//
+// It starts a new goroutine until either
+// a) the result channel of the wrapped next watcher is closed, or
+// b) Stop() was invoked on the returned watcher.
+func newDenylistWatch(l log.Logger, denylist map[string]struct{}, next watch.Interface) watch.Interface {
+	var (
+		result = make(chan watch.Event)
+		proxy  = watch.NewProxyWatcher(result)
+	)
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-next.ResultChan():
+				if !ok {
+					// stopping the wrapped next watcher is not necessary,
+					// as it closed the result channel itself.
+					return
```

Comment:
As upstream (next) closed its result channel, we have to close the downstream channel `result` too.

Reply:
fixed in the defer call: https://github.com/coreos/prometheus-operator/blob/02903b87b9d47559c1588e76359f979c959a1a94/pkg/listwatch/namespace_denylist.go#L129